### DISCUSSION TITLE
Allow DoesNotReturn methods in conditional expressions

### DISCUSSION THREAD

**Discussion by YohDeadfall** [dotnet/csharplang#3079] [Language Ideas] [answered]
For now the compiler cannot determine the type of a conditional expression when one of it's branches calls a method marked as `DoesNotReturn`. It isn't the right behavior since a non-returning method won't return control to the caller. Therefore, the method should not participate in type inference as when `throw` is used directly.

**Steps to reproduce**

```csharp
#nullable enable
private static string? Method(object? value) =>
    value is null? ThrowException() : value.ToString(); // error CS0173: Type of conditional expression cannot be determined because there is no implicit conversion between 'void' and 'string'

private static string? MethodWithoutError(object? value) =>
    value is null? throw new Exception() : value.ToString(); // compiled without issues

[DoesNotReturn]
private static void ThrowException() =>
    throw new Exception();
```

**Compiler version:** '3.3.1-beta4-19462-11 (66a912c9)'.
**Language version:** preview.

**Comment by 333fred** [comment, marked as answer]
> It isn't the right behavior since a non-returning method won't return control to the caller.

There is no guarantee of this, as the compiler does not enforce that this method actually does not return. For this, we'd need something like https://github.com/dotnet/csharplang/issues/538.

**Comment by jmarolf** [comment]
This makes sense to me since we don't want nullable attributes to change behavior of code. I would personally prefer a more full-fledged feature to support this like #538 instead of attributes changing type resolution behavior.
